### ISSUE TITLE
Function Loading Memory Leak

### ISSUE THREAD

**Issue by MicahSchiewe** [rune-rs/rune#273]
Hi all. A friend and I have been working on a project making heavy use of Rune, and have encountered some issues. While benchmarking our code using `jemalloc` to measure memory usage, we found that some scripts were leaking. We investigated the issue, and it seems that loading functions into the Rune VM causes memory leaks, with the issue affecting all the examples we benchmarked (including several copied from the repo and Rune book).

Example (using `jemallocator = "0.3.2"` and `jemalloc-ctl = "0.3.3"`):
Benchmark
```rust
use criterion::{black_box, criterion_group, criterion_main, Criterion};
use rune::{Diagnostics, Options, Sources};
use runestick::{Context, Module, Source};
use std::sync::Arc;

#[global_allocator]
static ALLOC: jemallocator::Jemalloc = jemallocator::Jemalloc;

fn rune_bench(source: &str) -> runestick::Result<()> {
    let mut context = Context::default();

    let module = Module::default();
    context.install(&module)?;

    let mut sources = Sources::new();
    sources.insert(Source::new("test", source));

    let mut diagnostics = Diagnostics::new();

    let unit = rune::load_sources(
        &context,
        &Options::default(),
        &mut sources,
        &mut diagnostics,
    )?;

    let vm = runestick::Vm::new(Arc::new(context.runtime()), Arc::new(unit));
    vm.execute(&["main"], ())?.complete()?;

    Ok(())
}

fn rune_benchmark_good(c: &mut Criterion) {
    let epoch = jemalloc_ctl::epoch::mib().unwrap();
    let allocated = jemalloc_ctl::stats::allocated::mib().unwrap();

    c.bench_function("Rune 1", |b| {
        b.iter(|| {
            epoch.advance().unwrap();
            let before = allocated.read().unwrap();

            black_box(rune_bench(
                r#"
            struct User {
            }
            
            impl User {
            }
            "#,
            ));

            epoch.advance().unwrap();
            let after = allocated.read().unwrap();
            println!("{} - {}", after, before);
        })
    });
}

fn rune_benchmark_bad(c: &mut Criterion) {
    let epoch = jemalloc_ctl::epoch::mib().unwrap();
    let allocated = jemalloc_ctl::stats::allocated::mib().unwrap();

    c.bench_function("Rune 2", |b| {
        b.iter(|| {
            epoch.advance().unwrap();
            let before = allocated.read().unwrap();

            black_box(rune_bench(r#"pub fn main() { }"#));

            epoch.advance().unwrap();
            let after = allocated.read().unwrap();
            println!("{} - {}", after, before);
        })
    });
}

criterion_group!(benches, rune_benchmark_good, rune_benchmark_bad);
criterion_main!(benches);
```
Dependencies
```toml
[dependencies]
rune = "0.9.0"
runestick = "0.9.0"

[dev-dependencies]
criterion = { version = "0.3.4", features = ["real_blackbox"] }
jemallocator = "0.3.2"
jemalloc-ctl = "0.3.3"
```
Do you think you could look into the issue, or help us figure out what we're doing wrong?

**Comment by udoprog** [comment]
I'm out right now but I'll give it a look. Thanks for the detailed repro!

**Comment by nnazo** [comment]
Just to help pinpoint things a bit -- it appears that it's coming from `rune::load_sources` whenever we have at least one `Source` in our `Sources` that has a function. You should be able to comment out the lines at the end creating and running the main function in the VM and it should still have the leak.

**Comment by udoprog** [comment]
So by using a [patched version of checkers](https://github.com/udoprog/checkers/pull/8), I've located the memory leak. Being fixed in #274. If anyone's interested, I refactored your test to this:

```
#[global_allocator]
static ALLOCATOR: checkers::Allocator = checkers::Allocator::system();

/* rune_bench impl */

fn main() {
    let snapshot = checkers::with(|| {
        let out = rune_bench("pub fn main() {}");
        assert!(out.is_ok());
    });

    checkers::verify!(snapshot);
}
```
